components/google_sheets/sources/common.js

```diff
@@ -0,0 +1,104 @@
+const { v4: uuid } = require("uuid");
+const google_sheets = require("../google_sheets.app");
+
+module.exports = {
+  props: {
+    google_sheets,
+    db: "$.service.db",
+    http: "$.interface.http",
+    timer: {
+      label: "Push notification renewal schedule",
+      description:
+        "The Google Drive API requires occasionally renewal of push notification subscriptions. **This runs in the background, so you should not need to modify this schedule**.",
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 60 * 6,
+      },
+    },
+    watchedDrive: { propDefinition: [google_sheets, "watchedDrive"] },
+  },
+  hooks: {
+    async activate() {
+      // Called when a component is created or updated. Handles all the logic
+      // for starting and stopping watch notifications tied to the desired files.
+
+      const channelID = this.db.get("channelID") || uuid();
+      const driveId = this.getDriveId();
+
+      const startPageToken = await this.google_sheets.getPageToken(driveId);
+      const { expiration, resourceId } = await this.google_sheets.watchDrive(
+        channelID,
+        this.http.endpoint,
+        startPageToken,
+        driveId,
+      );
+
+      // We use and increment the pageToken as new changes arrive, in run()
+      this.db.set("pageToken", startPageToken);
+
+      // Save metadata on the subscription so we can stop / renew later
+      // Subscriptions are tied to Google's resourceID, "an opaque value that
+      // identifies the watched resource". This value is included in request headers
+      this.db.set("subscription", { resourceId, expiration });
+      this.db.set("channelID", channelID);
+
+      await this.takeSheetSnapshot();
```

Comment:
Do you think this is worth moving to the `run` method (i.e. the first time the component runs)? I noticed the new updates source still took some time to run the `activate()` hook and I'm concerned even with the modified code, it might time out for the user.

Reply:
Yes, it's worth doing since the improvements don't really make any difference if the user selects every worksheet in a spreadsheet.

Reply:
Done!